### ISSUE TITLE
Advanced Machine Learning Model Development

### ISSUE THREAD

**Issue by edesz** [edesz/credit-card-churn#7]
- [x] model validation using training data
- [x] model evaluation using combined train+validation (train+val) data split and test data split

## Model Validation

The following steps should be implemented during model validation
1. perform the following using `prauc` as primary metric
    - model selection
    - feature selection
    - estimate model over-fitting
    - (optional) hyper-parameter tuning
2. perform threshold tuning in order to determine the best decision threshold using `f2_score` as the primary metric, for the best model determined from step 1
    - threshold tuning is important due to class imbalance
    - consider using `TunedThresholdClassifierCV` to perform automated threshold tuning ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TunedThresholdClassifierCV.html#sklearn.model_selection.TunedThresholdClassifierCV), [example](https://scikit-learn.org/stable/auto_examples/model_selection/plot_tuned_decision_threshold.html#sphx-glr-auto-examples-model-selection-plot-tuned-decision-threshold-py))
3. Best model found during step 1. and best decision threshold found during step 2. is trained on train+val data split and used to make probabilistic predictions (using `.pred_proba()`) of validation data split. For use in the next step, the probabilistic predictions should be stored in column named `y_pred_proba`.
2. Next, probabilistic predictions are converted into labels using the best decision threshold found in step 2. For use in next step, the labels should be stored in column named `y_pred`.

## Model Evaluation

1. Best model found during model validation is trained on train+val data split and used to make probabilistic predictions (using `.pred_proba()`) of test data split and train+val data split. For use in the next step, the probabilistic predictions should be stored in column named `y_pred_proba`.
2. Next, probabilistic predictions are converted into labels using the best decision threshold found during model evaluation. For use in next step, the labels should be stored in column named `y_pred`.
3. Next, model predictions on test split and train+val split are scored using `f2_score` as the primary metric.
4. Additional steps to consider
    - estimate and discuss impact of model over-fitting
    - get feature importances on predictions of test split using model's predictions and discuss

## Scoring Metrics

See the [project scope](https://github.com/edesz/credit-card-churn/blob/main/references/06_analysis.md#final-choice-of-machine-learning-metrics) for reasoning behind choice of scoring metrics for this use-case.

Below are example implementation of scoring metrics for use during model validation and model evaluation

1. for use in model validation ([link](https://machinelearningmastery.com/roc-curves-and-precision-recall-curves-for-classification-in-python/))
    ```python
    from sklearn.metrics import precision_recall_curve, auc
    # Calculate precision, recall, and thresholds
    precision, recall, _ = precision_recall_curve(y_true, y_pred_proba)
    # Calculate PR AUC using auc
    pr_auc = auc(recall, precision)
    ```
2. for use in model evaluation ([link](https://stackoverflow.com/a/67120544))
    ```python
    from sklearn.metrics import beta_score
    f2_score = beta_score(y_true, y_pred, beta=2, average='binary')
    ```

## Additional Notes

1. Consider using a pipeline object to perform feature transformation (normaliztion for numerical features, one-hot encoding for categorical features and ordinal encoding for ordinal features).
2. Consider using `class_weights='balanced'` to account for class imbalance
